tools/tplgfw-sanity-check.py

```diff
@@ -0,0 +1,134 @@
+
+import argparse
+import uuid
+from construct import this, ListContainer, Struct, Bytes, PaddedString
+from construct import Int16ul, Int32ul
+import tplgtool2
+
+class Fw_Parser:
+
+    def __init__(self):
+        self.fw_header_offset = 0x2000
+
+        self.Ext_header = Struct (
+            "id" / PaddedString(4, "ascii"),
+            "len" / Int32ul,
+            "version_major" / Int16ul,
+            "version_minor" / Int16ul,
+            "module_entries" / Int32ul,
+        )
+
+        self.Fw_binary_header = Struct (
+            "id" / PaddedString(4, "ascii"),
+            "len" / Int32ul,
+            "name" / PaddedString(8, "ascii"),
+            "preload_page_count" / Int32ul,
+            "flags" /Int32ul,
+            "feature_mask" / Int32ul,
+            "major" / Int16ul,
+            "minor" / Int16ul,
+            "hotfix" / Int16ul,
+            "build" / Int16ul,
+            "module_entries" / Int32ul,
+            "hw_buffer_add" / Int32ul,
+            "hw_buffer_len" / Int32ul,
+            "load_offset" / Int32ul,
+        )
+
+        self.Module_segment = Struct (
+            "flags" / Int32ul,
+            "v_base_addr" / Int32ul,
+            "file_offset" / Int32ul,
+        )
+
+        self.Module_entry = Struct (
+            "id" / Int32ul,
+            "name" / PaddedString(8, "ascii"),
+            "uuid" / Bytes(16),
+            "type" / Int32ul,
+            "hash" / Bytes(32),
+            "entry_point" / Int32ul,
+            "cfg_offset" / Int16ul,
+            "cfg_count" / Int16ul,
+            "affinity_mask" / Int32ul,
+            "instance_max_count" / Int16ul,
+            "stack_size" / Int16ul,
+            "segment" / self.Module_segment[3],
+        )
+
+        self.Ext_manifest = Struct (
+            "ext_header" / self.Ext_header,
+            "padding1" / Bytes(this.ext_header.len + self.fw_header_offset - self.Ext_header.sizeof()),
+            "fw_binary_header" / self.Fw_binary_header,
+            "modules" / self.Module_entry[this.fw_binary_header.module_entries],
+        )
+
+    def parse_file(self, filepath) -> ListContainer:
+        with open(filepath, "rb") as fw :
+            return self.Ext_manifest.parse_stream(fw)
+
+    def parse_fw_header(self, filepath) -> ListContainer:
+        with open(filepath, "rb") as fw :
+            return self.Ext_header.parse_stream(fw)
+
+if __name__ == "__main__":
+
+    def parse_cmdline():
+        parser = argparse.ArgumentParser(add_help=True, formatter_class=argparse.RawTextHelpFormatter,
+            description='A Python tool to match UUID in topology and FW, then find out whether all module in topoloy can be supported by this FW.\
+ Currently it only supports IPC4 FW and IPC3 & IPC4 topology. The output includes FW version and with an error message if they are not matched.\
+ And all the unsupported modules will be printed out.\
+ \n\nexample: tplgfw-sanity-check.py -t sof-tgl-nocodec.tplg -f basefw.bin')
```

Comment:
multi-line string with triple single quotes is a good practice, see https://github.com/thesofproject/sof/blob/fca84b8773ee4a8e6308a0842be6367be560453e/scripts/xtensa-build-zephyr.py#L175

Reply:
thanks for your advice!